system/jlib/jstring.cpp

```diff
@@ -937,28 +937,34 @@ StringBuffer & StringBuffer::replace(char oldChar, char newChar)
 // Copy source to result, replacing all occurrences of "oldStr" with "newStr"
 StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char *source, size_t lenOldStr, const char* oldStr, size_t lenNewStr, const char* newStr)
 {
-    if (lenSource)
-    {
-        size_t left = lenSource;
-        while (left >= lenOldStr)
+    if (lenSource >= lenOldStr)
+    {
+        // Reserve space when the new string is longer than the old string
+        unsigned steps = lenSource - lenOldStr + 1;
+        unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
+        result.ensureCapacity(maxResultLength);
```

Comment:
I'm not convinced this is worthwhile.  This could end up allocating a very large amount of memory, which is unlikely to be used.  The big benefit is to avoid reallocation from 0 up to the size of the source string which would have previously been the case.  I would use

results.ensureCapacity(lenSource);

The calculation is also incorrect it should be
 unsigned [max]steps = lenSource / lenOldStr;

Reply:
Changed to results.ensureCapacity(lenSource);